File: Python/LC/207.py
#! Re - J
from typing import List


class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:

        if len(prerequisites) == 0:
            return True

        graph = self.form_graph(prerequisites)

        if len(graph.keys()) > numCourses:
            return False

        for preq in prerequisites:
            start = preq[0]

            if self.find_cycle(graph, start, {}):
                return False

        return True

    @staticmethod
    def form_graph(vertices: List[List[int]]) -> dict:
        graph = {}

        for vertex in vertices:
            start = vertex[0]
            end = vertex[1]

            if start not in graph:
                graph[start] = [end]
            else:
                graph[start].append(end)

            if end not in graph:
                graph[end] = []

        return graph

    def find_cycle(self, graph: dict, vertex: int, visited: dict) -> bool:
        neighbours = graph[vertex] if vertex in graph else []

        if vertex in visited and visited[vertex]:
            return True

        visited[vertex] = True

        for neighbour in neighbours:
            if self.find_cycle(graph, neighbour, visited):
                return True

        visited[vertex] = False

        return False


if __name__ == '__main__':
    s = Solution()
    print(s.canFinish(2, [[1,0]]))
    print(s.canFinish(2, [[1,0],[0,1]]))
    print(s.canFinish(2, [[1,2],[2,3]]))
    print(s.canFinish(3, [[1,2],[2,3]]))
    print(s.canFinish(3, [[0,1],[0,2],[1,2]]))
    print(s.canFinish(4, [[2,0],[1,0],[3,1],[3,2],[1,3]]))

File: LC/207/207.java
class TopologicalSort {
    int count;
    int[] incomings;
    Map<Integer, List<Integer>> graph;
    
    public TopologicalSort(int count, int[][] prereq) {
        this.count = count;
        this.graph = new HashMap<>();
        this.incomings = new int[count];
        
        for (int[] pre : prereq) {
            int target = pre[0];
            int source = pre[1];
            
            if (!this.graph.containsKey(source)) this.graph.put(source, new ArrayList<Integer>());
            this.graph.get(source).add(target);
            this.incomings[target]++;
        }
    }
    
    public List<Integer> sort() {
        List<Integer> sorted = new ArrayList<>();
        Queue<Integer> queue = new LinkedList<>();
        for (int i = 0; i < this.incomings.length; i++) {
            if (this.incomings[i] == 0) queue.offer(i);
        }
        
        while (!queue.isEmpty()) {
            int node = queue.poll();
            sorted.add(node);
            
            List<Integer> neighbours = this.graph.getOrDefault(node, new ArrayList<Integer>());
            
            for (int neighbour : neighbours) {
                this.incomings[neighbour]--;
                if (this.incomings[neighbour] == 0) queue.offer(neighbour);
            }
        }
        
        return (sorted.size() != this.count) ? null : sorted;
    }
}

class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        if (numCourses == 0) return prerequisites.length == 0;
        TopologicalSort topo = new TopologicalSort(numCourses, prerequisites);
        return topo.sort() != null;
    }
}
